debezium-core/src/main/java/io/debezium/relational/TableSchemaBuilder.java

```diff
@@ -234,10 +235,13 @@ private void validateIncomingRowToInternalMetadata(int[] recordIndexes, Field[]
     protected Function<Object[], Struct> createValueGenerator(Schema schema, TableId tableId, List<Column> columns,
                                                               ColumnNameFilter filter, ColumnMappers mappers) {
         if (schema != null) {
-            int[] recordIndexes = indexesForColumns(columns);
-            Field[] fields = fieldsForColumns(schema, columns);
+            List<Column> columnsThatShouldBeAdded = columns.stream()
+                    .filter(column -> filter == null || filter.matches(tableId.catalog(), tableId.schema(), tableId.table(), column.name()))
+                    .collect(Collectors.toList());
+            int[] recordIndexes = indexesForColumns(columnsThatShouldBeAdded);
+            Field[] fields = fieldsForColumns(schema, columnsThatShouldBeAdded);
             int numFields = recordIndexes.length;
-            ValueConverter[] converters = convertersForColumns(schema, tableId, columns, filter, mappers);
+            ValueConverter[] converters = convertersForColumns(schema, tableId, columnsThatShouldBeAdded, filter, mappers);
```

Comment:
As you apply the filtering now already here, it should be removed from `convertersForColumns()`, also removing the `filter` parameter to that method.

Reply:
Done